### PR TITLE
【Infer Symbolic Shape No.125】【BUAA】Add BroadcastTensor, changed 3 files

### PR DESCRIPTION
**Pull request by Whsjrczr**
### PR Category

CINN

### PR Types

Others

### Description

加入broadcast_tensors:
    paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/multiary_infer_sym.cc
    paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/multiary_infer_sym.h
    paddle/phi/ops/yaml/ops.yaml
单测位于Paddle\test\legacy_test\test_broadcast_tensors_op.py，默认check_symbol_infer=True
本地单测结果如下：
屏幕截图 2024-09-25 214213

### CODE REVIEW

**Comment by gongshaotian** [comment]
旧PR：https://github.com/PaddlePaddle/Paddle/pull/67744

**Review by gongshaotian — APPROVED** [review]
LGTM

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/multiary_infer_sym.cc [hunk 1]
```diff
@@ -652,12 +652,40 @@ bool BilinearOpInferSymbolicShape(
...
+    pir::Operation *op, pir::InferSymbolicShapeContext *infer_context) {
+  const auto &input_shape_or_data_list =
+      infer_context->GetShapeOrDataForValue(op->operand_source(0))
+          .dyn_cast<symbol::TensorListShapeOrDataDimExprs>();
+  // 1. Find Output rank = max(Inputs rank)
+  size_t target_rank = 0;
+  for (const auto &input_shape_or_data : input_shape_or_data_list) {
+    size_t tmp_rank = input_shape_or_data.shape().size();
+    target_rank = std::max(int64_t(target_rank), int64_t(tmp_rank));
+  }
```
Comment (Hongqing-work) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/multiary_infer_sym.cc]: 可以减少点数据类型转换

**Review by Hongqing-work — DISMISSED** [review]

**Comment by gongshaotian** [comment]
已有单测，check_pir flag 已打开 test/legacy_test/test_broadcast_tensors_op.py 
image

**Comment by gongshaotian** [comment]
approved，coverage流水线超时取消了

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/multiary_infer_sym.cc [hunk 2]
```diff
@@ -652,12 +652,40 @@ bool BilinearOpInferSymbolicShape(
...
+  symbol::DimExprBuilder builder;
+  for (int i = 0; i < target_rank; i++) {
+    auto tmp_dim = symbol::DimExpr{1};
+    for (const auto &input_shape_or_data : input_shape_or_data_list) {
+      int axis = i - target_rank + input_shape_or_data.shape().size();
+      if (axis >= 0) {
+        infer_context->AddBroadcastableCstr(input_shape_or_data.shape()[axis],
+                                            tmp_dim);
+        tmp_dim = builder.Broadcast(input_shape_or_data.shape()[axis], tmp_dim);
+      }
```
Comment (zyfncg) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/multiary_infer_sym.cc]: image

int和size_t类型的运算要注意，会有越界的风险

Comment (Whsjrczr) [reply paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/multiary_infer_sym.cc]: 已修改

**Review by liym27 — APPROVED** [review]
approve for PR-CI-Auto-Paralle

**Review by zyfncg — APPROVED** [review]

**Pull request merged**